app/gui2/src/stores/project.ts

```diff
@@ -13,43 +38,87 @@ import { Awareness } from 'y-protocols/awareness'
  */
 export const useProjectStore = defineStore('project', () => {
   // inputs
-  const projectName = ref<string>()
   const observedFileName = ref<string>()
 
   const doc = new Y.Doc()
   const awareness = new Awareness(doc)
 
+  const config = useGuiConfig()
+  const projectId = config.value.startup?.project
+  if (projectId == null) throw new Error('Missing project ID')
+
+  const lsUrls = resolveLsUrl(config.value)
+
+  const rpcTransport = new WebSocketTransport(lsUrls.rpcUrl)
+  const rpcRequestManager = new RequestManager([rpcTransport])
+  const rpcClient = new Client(rpcRequestManager)
+  const lsRpcConnection = new LanguageServer(rpcClient)
+
+  const undoManager = new Y.UndoManager([], { doc })
+
+  useWindowEvent('keydown', (e) => {
+    if (modKey(e) && e.key === 'z') {
+      console.log('undo')
+      undoManager.undo()
+    } else if (modKey(e) && e.key === 'y') {
+      console.log('redo')
+      undoManager.redo()
+    }
+  })
```

Comment:
fwiw for the dashboard I've implemented a global shortcut registry that we could potentially refactor out, so that both gui and dashboard could use their own instance of it